platforms/ide/tooling-api/src/main/java/org/gradle/tooling/internal/adapter/ProtocolToModelAdapter.java

```diff
@@ -243,6 +230,25 @@ public Iterable<? extends Class<?>> successors(Class<?> node) {
         return filteredImplementedInterfaces;
     }
 
+    private static <T> Set<Class<?>> walkTypeHierarchyAndExtractInterfaces(Class<?> clazz) {
```

Comment:
💭  Ideally, we want to reuse existing type-walking code from `Types` utils. We don't have to refactor it here, but I think we at least need a TODO, so it's possible to find semi-duplicates like these.

Reply:
I did reuse it first, but it doesn't work, because those modules are compiled with Java 8 and the code here needs to be Java 7 compatible. So when you reuse Types you have the exact same problem as the original one we're fixing. Now, maybe there is a better solution than this half-duplication, I'm not saying that, just that there was a reason for it.

Reply:
I hope that this is solvable by extracting a new Gradle module with utilities aimed at Java 6 level boilerplate. Then those should be usable from all places, including TAPI.